plugins/org.python.pydev/src/org/python/pydev/editor/correctionassist/heuristics/AssistFString.java

```diff
@@ -0,0 +1,132 @@
+            IPythonNature nature, IPyEdit edit, int offset) throws BadLocationException, MisconfigurationException {
+        PySourceViewer viewer = null;
+        if (edit instanceof PyEdit) { //only in tests it is actually null
+            viewer = ((PyEdit) edit).getPySourceViewer();
+        }
+        List<ICompletionProposalHandle> lst = new ArrayList<ICompletionProposalHandle>();
+
+        IDocument doc = ps.getDoc();
+        ITypedRegion partition = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc)).getPartition(offset);
+
+        FastStringBuffer strBuf = new FastStringBuffer();
+        strBuf.append('f').append(doc.get(partition.getOffset(), partition.getLength()));
+
+        FastStringBuffer variablesBuf = new FastStringBuffer();
+        // get string end offset
+        int partitionEndOffset = partition.getOffset() + partition.getLength();
+        int i = partitionEndOffset;
+        try {
+            // get format end offset
+            ParsingUtils parsingUtils = ParsingUtils.create(doc);
+            while (i < parsingUtils.len()) {
+                char c = parsingUtils.charAt(i);
+                if (Character.isWhitespace(c)) {
+                    i++;
+                    continue;
+                }
+                if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
+                    i = parsingUtils.eatPar(i, null, c) + 1;
+                    break;
+                }
+                if (Character.isJavaIdentifierPart(c)) {
+                    i++;
+                    while (i < parsingUtils.len()) {
+                        c = parsingUtils.charAt(i);
+                        if (Character.isJavaIdentifierPart(c) || c == '.') {
+                        } else {
+                            if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
+                                i = parsingUtils.eatPar(i, null, c);
+                            }
+                            break; // inner break
+                        }
+                        i++;
+                    }
+                    continue;
+                }
+                i++;
+            }
+
+            variablesBuf.append(doc.get(partitionEndOffset, i - partitionEndOffset));
+            // get only the variable names and remove the last comma
+            variablesBuf.trim().deleteFirst();
+            if (variablesBuf.trim().startsWith('(')) {
+                variablesBuf.deleteFirst();
+                variablesBuf.deleteLast();
+                variablesBuf.trim();
+            }
+            if (variablesBuf.endsWith(',')) {
+                variablesBuf.deleteLast();
+                variablesBuf.trim();
+            }
+
+            // get variables without literals
+            String variablesWithoutLiterals = PySelection.getLineWithoutCommentsOrLiterals(variablesBuf.toString());
+            // get variable by variable and edit the f-string output
```

Comment:
You should be able to do: `StringUtils.split(variablesWithoutLiterals, ',')` to get all the contents, then knowing how many are there you can see if it matches the number of `%s` strings found -- so you can properly handle the case where there's only one `%s` but multiple variables.

i.e.: doing something as `"%s - %s" % a"` should give as a result `f"{a[0]} - {a[1]}"`.

Reply:
The problem with this is if there is a dict in the variables buf, the split will take the commas that separates the keys
i.e.: `"d = %s" % {'id': 1, 'something': 10}` this would return size == 2

Reply:
A dict isn't valid for formatting with `%s`, so, thinking about it now, we should not support `{`, only `(` and `[`.